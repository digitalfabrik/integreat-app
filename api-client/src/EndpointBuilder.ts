import Endpoint from './Endpoint'
import { MapResponseType } from './MapResponseType'
import { MapParamsToUrlType } from './MapParamsToUrlType'
import { MapParamsToBodyType } from './MapParamsToBody'
/**
 * Helper class to build a {@link Endpoint}
 */

class EndpointBuilder<P, T> {
  name: string
  paramsToBodyMapper: MapParamsToBodyType<P> | null | undefined
  paramsToUrlMapper: MapParamsToUrlType<P>
  mapper: MapResponseType<P, T>
  responseOverride: T | null | undefined
  errorOverride: Error | null | undefined

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor(name: string) {
    this.name = name
  }

  /**
   * Adds a state to url mapper to the builder
   * @param paramsToUrlMapper The paramsToUrlMapper which is mapping the params to a url
   * @return {EndpointBuilder} The builder itself
   */
  withParamsToUrlMapper(paramsToUrlMapper: MapParamsToUrlType<P>): EndpointBuilder<P, T> {
    this.paramsToUrlMapper = paramsToUrlMapper
    return this
  }

  withParamsToBodyMapper(paramsToBodyMapper: MapParamsToBodyType<P> | null | undefined): EndpointBuilder<P, T> {
    this.paramsToBodyMapper = paramsToBodyMapper
    return this
  }

  /**
   * Adds a json mapper to the builder
   * @param mapper The mapper which maps json from our cms to models
   * @return {EndpointBuilder} The builder itself
   */
  withMapper(mapper: MapResponseType<P, T>): EndpointBuilder<P, T> {
    this.mapper = mapper
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride(responseOverride: T): EndpointBuilder<P, T> {
    this.responseOverride = responseOverride
    return this
  }

  /**
   * Fetcher throws an error. Useful for testing.
   * @param errorOverride {*} The error
   * @return {EndpointBuilder} The builder itself
   */
  withErrorOverride(errorOverride: Error): EndpointBuilder<P, T> {
    this.errorOverride = errorOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build(): Endpoint<P, T> {
    if (!this.name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this.paramsToUrlMapper) {
      throw Error('You have to set a url mapper to build an endpoint!')
    }

    if (!this.mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    return new Endpoint<P, T>(
      this.name,
      this.paramsToUrlMapper,
      this.paramsToBodyMapper,
      this.mapper,
      this.responseOverride,
      this.errorOverride
    )
  }
}

export default EndpointBuilder
