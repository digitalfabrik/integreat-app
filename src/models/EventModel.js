// @flow

import type Moment from 'moment'
import LocationModel from './LocationModel'
import DateModel from './DateModel'
import ExtendedPageModel from './ExtendedPageModel'
import FeaturedImageModel from './FeaturedImageModel'

class EventModel extends ExtendedPageModel {
  _date: DateModel
  _location: LocationModel
  _excerpt: string
  _featuredImage: ?FeaturedImageModel

  constructor (params: {| path: string, title: string, content: string, thumbnail: string,
    date: DateModel, location: LocationModel, excerpt: string, availableLanguages: Map<string, string>,
    lastUpdate: Moment, hash: string, featuredImage: ?FeaturedImageModel |}) {
    const {date, location, excerpt, featuredImage, ...other} = params
    super(other)
    this._date = date
    this._location = location
    this._excerpt = excerpt
    this._featuredImage = featuredImage
  }

  get date (): DateModel {
    return this._date
  }

  get location (): LocationModel {
    return this._location
  }

  get excerpt (): string {
    return this._excerpt
  }

  get featuredImage (): ?FeaturedImageModel {
    return this._featuredImage
  }
}

export default EventModel
