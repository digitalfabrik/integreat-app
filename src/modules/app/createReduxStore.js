// @flow

import type { Store } from 'redux'
import { applyMiddleware, createStore } from 'redux'
import { createLogger } from 'redux-logger'
import { AsyncStorage } from 'react-native'

import { languageReducer, uiDirectionReducer } from 'modules/i18n/reducers'
import toggleDarkModeReducer from '../theme/reducers'
import {
  checkInternetConnection,
  createNetworkMiddleware,
  networkEventsListenerSaga,
  offlineActionTypes,
  reducer as reactNativeOfflineReducer
} from 'react-native-offline'
import type { Saga } from 'redux-saga'
import createSagaMiddleware from 'redux-saga'
import { all, fork } from 'redux-saga/effects'
import { persistCombineReducers, persistStore } from 'redux-persist'
import type { PersistConfig, Persistor } from 'redux-persist/src/types'
import type { StateType } from './StateType'
import type { StoreActionType } from './StoreActionType'
import fetchCities from '../endpoint/sagas/fetchCities'
import fetchCategories from '../endpoint/sagas/fetchCategories'
import hardSet from 'redux-persist/lib/stateReconciler/hardSet'
import categoriesReducer from '../endpoint/reducers/categoriesReducer'
import fileCacheReducer from '../endpoint/reducers/fileCacheReducer'
import citiesReducer from '../endpoint/reducers/cititesReducer'
import languagesReducer from '../endpoint/reducers/languagesReducer'
import currentCityReducer from '../categories/reducers/currentCityReducer'
import { composeWithDevTools } from 'redux-devtools-extension'

function * rootSaga (): Saga<void> {
  yield all([
    fork(fetchCities),
    fork(fetchCategories),
    fork(networkEventsListenerSaga, {})
  ])
}

const createReduxStore = (callback: () => void): { store: Store<StateType, StoreActionType>, persistor: Persistor } => {
  const sagaMiddleware = createSagaMiddleware()

  const initialState: StateType = {
    uiDirection: 'ltr',
    language: 'en',
    currentCity: null,
    darkMode: false,

    cities: {json: null, error: null},
    categories: {},
    languages: {},
    fileCache: {},

    network: {isConnected: false, actionQueue: []}
  }

  const persistConfig: PersistConfig = {
    version: 1,
    key: 'root',
    storage: AsyncStorage,
    stateReconciler: hardSet
  }

  // Create this reducer only once. It is not pure!
  const persitedReducer = persistCombineReducers(persistConfig, {
    uiDirection: uiDirectionReducer,
    language: languageReducer,
    currentCity: currentCityReducer,
    darkMode: toggleDarkModeReducer,

    cities: citiesReducer,
    categories: categoriesReducer,
    languages: languagesReducer,
    fileCache: fileCacheReducer,

    network: reactNativeOfflineReducer
  })

  const rootReducer = (state, action) => {
    if (!state) {
      return initialState
    }
    return persitedReducer(state, action)
  }

  // TODO: Disable logger and dev tools in production

  const middleware = applyMiddleware(createNetworkMiddleware(), sagaMiddleware, createLogger())

  const store = createStore(rootReducer, initialState, composeWithDevTools(middleware))

  const persistor = persistStore(
    store,
    undefined,
    async () => {
      const isConnected: boolean = await checkInternetConnection()
      store.dispatch({
        type: offlineActionTypes.CONNECTION_CHANGE,
        payload: isConnected
      })
      sagaMiddleware.run(rootSaga)
      callback()
    }
  )

  return {store, persistor}
}

export default createReduxStore
