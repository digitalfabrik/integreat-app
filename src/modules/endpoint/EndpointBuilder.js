// @flow

import Endpoint from './Endpoint'
import type { MapResponseType } from './MapResponseType'
import type { MapParamsToUrlType } from './MapParamsToUrlType'
import type { MapParamsToBodyType } from './MapParamsToBody'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder<P, T> {
  name: string
  paramsToBodyMapper: ?MapParamsToBodyType<P>
  paramsToUrlMapper: MapParamsToUrlType<P>
  mapper: MapResponseType<P, T>
  responseOverride: ?T
  errorOverride: ?Error

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name: string) {
    this.name = name
  }

  /**
   * Adds a state to url mapper to the builder
   * @param paramsToUrlMapper The paramsToUrlMapper which is mapping the params to a url
   * @return {EndpointBuilder} The builder itself
   */
  withParamsToUrlMapper (paramsToUrlMapper: MapParamsToUrlType<P>): EndpointBuilder<P, T> {
    this.paramsToUrlMapper = paramsToUrlMapper
    return this
  }

  withParamsToBodyMapper (paramsToBodyMapper: ?MapParamsToBodyType<P>): EndpointBuilder<P, T> {
    this.paramsToBodyMapper = paramsToBodyMapper
    return this
  }

  /**
   * Adds a json mapper to the builder
   * @param mapper The mapper which maps json from our cms to models
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper: MapResponseType<P, T>): EndpointBuilder<P, T> {
    this.mapper = mapper
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride (responseOverride: T): EndpointBuilder<P, T> {
    this.responseOverride = responseOverride
    return this
  }

  /**
   * Fetcher throws an error. Useful for testing.
   * @param errorOverride {*} The error
   * @return {EndpointBuilder} The builder itself
   */
  withErrorOverride (errorOverride: Error): EndpointBuilder<P, T> {
    this.errorOverride = errorOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build (): Endpoint<P, T> {
    if (!this.name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this.paramsToUrlMapper) {
      throw Error('You have to set a url mapper to build an endpoint!')
    }

    if (!this.mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    return new Endpoint<P, T>(
      this.name,
      this.paramsToUrlMapper, this.paramsToBodyMapper, this.mapper,
      this.responseOverride, this.errorOverride
    )
  }
}

export default EndpointBuilder
