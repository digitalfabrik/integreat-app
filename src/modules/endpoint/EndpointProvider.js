import React from 'react'
import PropTypes from 'prop-types'
import Endpoint from './Endpoint'

class EndpointProvider extends React.Component {
  static propTypes = {
    endpoints: PropTypes.arrayOf(PropTypes.instanceOf(Endpoint)).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    getEndpoint: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      getEndpoint: endpointName => {
        const endpoint = this.endpoints[endpointName]
        if (!endpoint) {
          throw new Error(`The endpoint ${endpointName} was not found. Maybe you forgot to register the endpoint!`)
        }
        return endpoint
      }
    }
  }

  constructor (props, context) {
    super(props, context)
    this.endpoints = props.endpoints.reduce((accumulator, endpoint) => {
      accumulator[endpoint.stateName] = endpoint
      return accumulator
    }, {})
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
