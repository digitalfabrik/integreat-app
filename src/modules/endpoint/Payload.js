// @flow

import isUrl from 'is-url'

/**
 * The payload gets stored in the redux store and holds the information about a fetch
 */
class Payload<T> {
  _isFetching: boolean
  _data: ?T
  _error: ?Error
  _requestUrl: ?string
  _fetchDate: number

  constructor (
    isFetching: boolean,
    requestUrl: ?string = null,
    data: ?T = null,
    error: ?Error = null,
    fetchDate: number = Date.now()
  ) {
    this._isFetching = isFetching
    this._fetchDate = fetchDate
    this._error = error
    this._requestUrl = requestUrl
    this._data = data

    if (requestUrl) {
      if (!isUrl(requestUrl)) {
        throw new Error('requestUrl must be a valid URL')
      }
    }

    if (error && data) {
      throw new Error('data and error can not be set at the same time')
    }
  }

  get fetchDate (): number {
    return this._fetchDate
  }

  get isFetching (): boolean {
    return this._isFetching
  }

  get data (): ?T {
    return this._data
  }

  get error (): ?Error {
    return this._error
  }

  get requestUrl (): ?string {
    return this._requestUrl
  }
}

export default Payload
