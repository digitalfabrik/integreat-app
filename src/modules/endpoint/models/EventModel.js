// @flow

import type Moment from 'moment'
import LocationModel from './LocationModel'
import DateModel from './DateModel'
import PageModel from './PageModel'

class EventModel extends PageModel {
  _date: DateModel
  _location: LocationModel
  _excerpt: string

  constructor (params: {|id: number, path: string, title: string, content: string, thumbnail: string,
    date: DateModel, location: LocationModel, excerpt: string, availableLanguages: Map<string, string>,
    lastUpdate: Moment|}) {
    const {date, location, excerpt, ...other} = params
    super(other)
    this._date = date
    this._location = location
    this._excerpt = excerpt
  }

  get date (): DateModel {
    return this._date
  }

  get location (): LocationModel {
    return this._location
  }

  get excerpt (): string {
    return this._excerpt
  }
}

export default EventModel
