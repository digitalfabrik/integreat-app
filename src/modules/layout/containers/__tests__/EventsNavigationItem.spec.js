import React from 'react'
import { Provider } from 'react-redux'
import EndpointBuilder from 'modules/endpoint/EndpointBuilder'
import { mount, shallow } from 'enzyme'
import ConnectedEventsNavigationItem, { EventsNavigationItem } from '../EventsNavigationItem'
import EndpointProvider from '../../../endpoint/EndpointProvider'
import createReduxStore from '../../../app/createReduxStore'
import createHistory from '../../../app/createHistory'
import Payload from '../../../endpoint/Payload'
import EventModel from '../../../endpoint/models/EventModel'
import moment from 'moment-timezone'

describe('EventsNavigationItem', () => {
  const language = 'de'
  const href = 'href1'
  const text = 'text1'
  const tooltip = 'tooltip1'
  const selected = false

  const events = [
    new EventModel({
      id: 1234,
      title: 'first Event',
      availableLanguages: {de: '1235', ar: '1236'},
      startDate: moment.tz('2017-11-18 09:30:00', 'UTC'),
      endDate: moment.tz('2017-11-18 19:30:00', 'UTC'),
      allDay: true
    }),
    new EventModel({
      id: 2,
      title: 'second Event',
      startDate: moment.tz('2017-11-18 09:30:00', 'UTC'),
      endDate: moment.tz('2017-11-18 19:30:00', 'UTC'),
      allDay: true
    })]

  it('should render active if events are available', () => {
    const component = shallow(
      <EventsNavigationItem href={href} text={href} tooltip={tooltip} selected={selected} events={events} />)
    expect(component).toMatchSnapshot()
  })

  it('should render inactive if events are undefined or empty', () => {
    const component = shallow(
      <EventsNavigationItem href='href1' text='text1' tooltip='tooltip1' selected={selected} events={undefined} />)
    expect(component).toMatchSnapshot()
    component.setProps({...component.props(), events: []})
    expect(component).toMatchSnapshot()
  })

  describe('connect()', () => {
    const location = 'augsburg'
    const route = '/:location/:language'

    const createComponent = (fetching = false, failure = false) => {
      const eventsEndpoint = new EndpointBuilder('events')
        .withStateToUrlMapper(() => 'https://weird-endpoint/api.json')
        .withMapper(json => json)
        .withResponseOverride(events)
        .withErrorOverride(failure ? 'Error No. 5' : undefined)
        .build()

      const store = createReduxStore(createHistory, {
        events: new Payload(fetching),
        router: {params: {location, language}, route}
      })
      return mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[eventsEndpoint]}>
            <ConnectedEventsNavigationItem href='href1' text='text1' tooltip='tooltip1' selected={false} />
          </EndpointProvider>
        </Provider>
      )
    }

    it('should be rendered if still fetching', () => {
      const navigationItem = createComponent(true).find(EventsNavigationItem)
      expect(navigationItem.props()).toEqual({events: undefined, href, selected, text, tooltip})
    })

    it('should be rendered if an error occured', () => {
      const navigationItem = createComponent(false, true).find(EventsNavigationItem)
      expect(navigationItem.props()).toEqual({events: undefined, href, selected, text, tooltip, error: 'Error No. 5'})
    })

    it('should be rendered if events are available', () => {
      const navigationItem = createComponent(false, false).find(EventsNavigationItem)
      expect(navigationItem.props()).toEqual({events, href, selected, text, tooltip})
    })
  })
})
