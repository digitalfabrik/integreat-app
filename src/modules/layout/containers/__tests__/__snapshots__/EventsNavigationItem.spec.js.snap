// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsNavigationItem should render active if events are available 1`] = `
<HeaderNavigationItem
  active={true}
  href="href1"
  selected={false}
  text="href1"
  tooltip="tooltip1"
/>
`;

exports[`EventsNavigationItem should render inactive if events are undefined or empty 1`] = `
<HeaderNavigationItem
  active={false}
  href="href1"
  selected={false}
  text="text1"
  tooltip="tooltip1"
/>
`;

exports[`EventsNavigationItem should render inactive if events are undefined or empty 2`] = `
<HeaderNavigationItem
  active={false}
  href="href1"
  selected={false}
  text="text1"
  tooltip="tooltip1"
/>
`;
